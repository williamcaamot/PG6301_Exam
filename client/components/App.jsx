import React, {createContext, useEffect, useState} from "react";
import {BrowserRouter, Route, Router, Routes} from "react-router-dom";
import Login from "./login/Login.jsx";
import Header from "./Header.jsx";
import Home from "./home/Home.jsx";
import GoogleLoginCallback from "./login/GoogleLoginCallback.jsx";
import EntraIdLoginCallback from "./login/EntraIdLoginCallback.jsx";
import Chat from "./chat/Chat.jsx";
import Profile from "./profile/Profile.jsx";


export const AppContext = createContext();

export function App() {

    const [user, setUser] = useState(null);

    async function fetchUser() {
        const res = await fetch("/api/v1/login");
        if (res.status === 401) {
            setUser(undefined);

        } else {
            const user = await res.json();
            setUser(user);
            console.log(user)
        }
    }

    useEffect(() => {
        fetchUser();
    }, []);


    return <AppContext.Provider value={{user, setUser}}>
        <BrowserRouter>
        <Header/>
        <Routes>
            <Route path={"/login/callback/entraid"} element={<EntraIdLoginCallback/>}/>
            <Route path={"login/callback/google"} element={<GoogleLoginCallback/>}/>
            <Route path={"/login"} element={<Login/>}/>
            <Route path={"/profile"} element={<Profile/>}/>
            <Route path={"/chat"} element={<Chat/>}/>
            <Route path={"/"} element={<Home/>}/>
        </Routes>


    </BrowserRouter>
    </AppContext.Provider>
}