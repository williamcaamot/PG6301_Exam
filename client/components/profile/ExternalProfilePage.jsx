import React, { useEffect, useState } from "react";
import ErrorMessage from "../globals/ErrorMessage.jsx";
import { useMatch } from "react-router-dom";
import Profile from "./Profile.jsx";

function ExternalProfilePage() {
  const [errorMessage, setErrorMessage] = useState();

  const match = useMatch("/profile/:email");
  const email = match.params.email;

  const [user, setUser] = useState();

  async function getUser() {
    try {
      const res = await fetch(`/api/v1/profile/${email}`);
      const { message, data } = await res.json();

      if (res.status !== 200) {
        setErrorMessage(message || "En ukjent feil har oppstÃ¥tt");
        return;
      }
      setUser(data);
    } catch (e) {
      setErrorMessage(e.message);
    }
  }
  useEffect(() => {
    getUser();
  }, []);

  return (
    <>
      <div className={"pageContentWrapper"}>
        <div className={"innerWrapper"}>
          <ErrorMessage message={errorMessage} />
          <Profile user={user} />
        </div>
      </div>
    </>
  );
}

export default ExternalProfilePage;
