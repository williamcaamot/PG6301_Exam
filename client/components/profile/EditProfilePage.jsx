import React, { useContext, useEffect, useState } from "react";
import { AppContext } from "../App.jsx";
import Profile from "./Profile.jsx";
import SuccessMessage from "../globals/SuccessMessage.jsx";
import ErrorMessage from "../globals/ErrorMessage.jsx";
import { Link, useNavigate } from "react-router-dom";
import { setTimeout } from "stream-http/lib/request.js";

function EditProfilePage() {
  const [successMessage, setSuccessMessage] = useState();
  const [errorMessage, setErrorMessage] = useState();

  const { user, setUser } = useContext(AppContext);

  const [nickname, setNickname] = useState();
  const [bio, setBio] = useState();

  const navigate = useNavigate();

  useEffect(() => {
    if (user) {
      setNickname(user.nickname);
      setBio(user.bio);
    }
  }, [user]);

  async function handleUpdateUser(event) {
    event.preventDefault();
    const newUserDetails = { nickname: nickname, bio: bio };

    let res = await fetch(`/api/v1/profile/${user.email}`, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(newUserDetails),
    });
    const { message, data } = await res.json();
    if (res.status === 201) {
      console.log(data);
      setUser(data);
      setSuccessMessage(message);
      setErrorMessage(null);
      navigate("/profile");
    } else {
      setErrorMessage(message);
    }
  }

  if (!user) {
    return (
      <div className={"pageContentWrapper"}>
        <div className={"innerWrapper"}>
          <h2>Du må være logget inn for å kunne redigere profilen sin</h2>
        </div>
      </div>
    );
  }
  return (
    <>
      <div className={"pageContentWrapper"}>
        <div className={"innerWrapper"}>
          <form onSubmit={handleUpdateUser}>
            <p>
              Nickname:{" "}
              <input
                value={nickname}
                onInput={(event) => setNickname(event.target.value)}
              />
            </p>
            <p>
              Bio:{" "}
              <input
                value={bio}
                onInput={(event) => setBio(event.target.value)}
              />
            </p>
            <button>Oppdater</button>
            <Link to={"/profile"}>
              <button>Avbryt, gå tilabke</button>
            </Link>
          </form>

          <SuccessMessage message={successMessage} />
          <ErrorMessage message={errorMessage} />
        </div>
      </div>
    </>
  );
}

export default EditProfilePage;
