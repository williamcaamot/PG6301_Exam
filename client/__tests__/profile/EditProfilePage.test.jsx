import renderer, { act } from "react-test-renderer";
import { MemoryRouter } from "react-router-dom";
import ExternalProfilePage from "../../components/profile/ExternalProfilePage.jsx";
import EditProfilePage from "../../components/profile/EditProfilePage.jsx";
import { App, AppContext } from "../../components/App.jsx";

describe("edit user component", () => {
  it("should render the edit user component", async () => {
    const user = {
      _id: "655dd38a4c15f00c20bc3fcf",
      bio: "dawawddawJeg er kul",
      email: "williamcoucheronaamot@gmail.com",
      family_name: "Coucheron-Aamot",
      name: "William",
      nickname: "WILLIAM123wwwww",
      picture:
        "https://lh3.googleusercontent.com/a/ACg8ocIOgSmXkWTJqoKPe5Hk3JlkkhfUNer4sWkVo3UuIa_dzg=s96-c",
    };

    async function setUser() {}

    let component;
    await act(async () => {
      component = renderer.create(
        <AppContext.Provider value={{ user, setUser }}>
          <MemoryRouter initialEntries={["/profile/william@gmail.com"]}>
            <EditProfilePage handleUpdateUser={() => {}} />
          </MemoryRouter>
        </AppContext.Provider>,
      );
    });
    expect(component).toMatchSnapshot();
  });

  it("should must be signed in to edit", async () => {
    const user = null;

    async function setUser() {}

    let component;
    await act(async () => {
      component = renderer.create(
        <AppContext.Provider value={{ user, setUser }}>
          <MemoryRouter initialEntries={["/profile/william@gmail.com"]}>
            <EditProfilePage handleUpdateUser={() => {}} />
          </MemoryRouter>
        </AppContext.Provider>,
      );
    });
    expect(component).toMatchSnapshot();
  });
});
