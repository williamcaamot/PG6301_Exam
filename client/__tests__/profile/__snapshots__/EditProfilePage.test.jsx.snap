// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edit user component should must be signed in to edit 1`] = `
<div
  className="pageContentWrapper"
>
  <div
    className="innerWrapper"
  >
    <h2>
      Du må være logget inn for å kunne redigere profilen sin
    </h2>
  </div>
</div>
`;

exports[`edit user component should render the edit user component 1`] = `
<div
  className="pageContentWrapper"
>
  <div
    className="innerWrapper"
  >
    <form
      onSubmit={[Function]}
    >
      <p>
        Nickname: 
        <input
          onInput={[Function]}
          value="WILLIAM123wwwww"
        />
      </p>
      <p>
        Bio: 
        <input
          onInput={[Function]}
          value="dawawddawJeg er kul"
        />
      </p>
      <button>
        Oppdater
      </button>
      <a
        href="/profile"
        onClick={[Function]}
      >
        <button>
          Avbryt, gå tilabke
        </button>
      </a>
    </form>
  </div>
</div>
`;
